### PR TITLE
[CSS Shape function] Add parsing support and storage for shape()

### PR DESCRIPTION
**Pull request by smfr**
#### d2911575fae37f3a740a934d2c2f874b94dd61c3
<pre>
[CSS Shape function] Add parsing support and storage for shape()
<a href="https://bugs.webkit.org/show_bug.cgi?id=277347">https://bugs.webkit.org/show_bug.cgi?id=277347</a>
<a href="https://rdar.apple.com/132814728">rdar://132814728</a>

Reviewed by Tim Nguyen.

Implement parsing, property value and computed style support for the `shape()` function[1]. We support the syntax
as of the most recent edit[2] but without the flexible order of points and control points due to [3].

Added BasicShapesShapeSegmentConversion.h/cpp with helpers to convert between CSS values and the internal shapes.
CSSShapeValue::customCSSText() is implemented.

BasicShapeShape stores its segments as a vector of std::variant&lt;&gt;, much as we do for Paths, which avoids heap allocations
per segment.

Plumb Style::BuilderState deeper into BasicShapeConversion functions. This required passing an optional zoom override,
used in some SVG cases (exercised by layout tests).

Tentative WPT are included.

Based on code by Noam Rosenthal.

[1] <a href="https://drafts.csswg.org/css-shapes-2/#shape-function">https://drafts.csswg.org/css-shapes-2/#shape-function</a>
[2] <a href="https://github.com/w3c/csswg-drafts/commit/9770805c4e53">https://github.com/w3c/csswg-drafts/commit/9770805c4e53</a>
[3] <a href="https://github.com/w3c/csswg-drafts/issues/10666">https://github.com/w3c/csswg-drafts/issues/10666</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/animations/clip-path-interpolation-shape-expected.txt: Interpolation is not yet supported.
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-shape-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-computed.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-computed.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-invalid.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-invalid.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-valid.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-valid.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-path-interpolation-008-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/BasicShapeConversion.cpp:
(WebCore::valueForBasicShape):
(WebCore::convertToLengthPoint):
(WebCore::basicShapeForValue):
(WebCore::basicShapePathForValue):
(WebCore::basicShapeShapeForValue):
* Source/WebCore/css/BasicShapeConversion.h:
* Source/WebCore/css/BasicShapesShapeSegmentConversion.cpp: Added.
(WebCore::lengthToCSSValue):
(WebCore::lengthPointToCSSValue):
(WebCore::lengthSizeToCSSValue):
(WebCore::toCSSShapeSegmentValue):
(WebCore::fromCSSShapeSegmentValue):
* Source/WebCore/css/BasicShapesShapeSegmentConversion.h: Copied from Source/WebCore/css/BasicShapeConversion.h.
* Source/WebCore/css/CSSBasicShapes.cpp:
(WebCore::CSSShapeValue::CSSShapeValue):
(WebCore::CSSShapeValue::customCSSText const):
(WebCore::CSSShapeValue::equals const):
* Source/WebCore/css/CSSShapeSegmentValue.cpp:
(WebCore::CSSShapeSegmentValue::customCSSText const):
(WebCore::CSSShapeSegmentValue::toShapeSegment const):
* Source/WebCore/css/CSSShapeSegmentValue.h:
* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::convertToLengthSize):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBasicShapePolygon):
(WebCore::CSSPropertyParserHelpers::consumeBasicShapePath):
(WebCore::CSSPropertyParserHelpers::consumeCoordinatePair):
(WebCore::CSSPropertyParserHelpers::consumeShapeCommand):
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeShape):
(WebCore::CSSPropertyParserHelpers::consumeBasicShape):
* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/BasicShapes.h:
* Source/WebCore/rendering/style/BasicShapesShape.cpp: Added.
(WebCore::BasicShapeShape::create):
(WebCore::BasicShapeShape::BasicShapeShape):
(WebCore::BasicShapeShape::clone const):
(WebCore::BasicShapeShape::path const):
(WebCore::BasicShapeShape::canBlend const):
(WebCore::BasicShapeShape::blend const):
(WebCore::BasicShapeShape::operator== const):
(WebCore::BasicShapeShape::dump const):
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/BasicShapesShape.h: Added.
(WebCore::ShapeSegmentBase::ShapeSegmentBase):
(WebCore::ShapeSegmentBase::affinity const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertSVGPath):
(WebCore::Style::BuilderConverter::convertPathOperation):
(WebCore::Style::BuilderConverter::convertShapeValue):
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::style const):

Canonical link: <a href="https://commits.webkit.org/281805@main">https://commits.webkit.org/281805@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/6b1f92109b293930a2901287eb75d8a339a54718

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61030 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40391 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13611 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64966 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11579 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63160 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48068 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11854 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49330 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8039 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/63064 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37587 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52870 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30160 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34269 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10098 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10491 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/54134 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56090 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10393 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66697 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60278 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4977 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10209 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56700 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4999 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52837 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56890 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13619 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4122 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/82033 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36195 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14294 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37278 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38372 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37022 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/css/CSSShapeSegmentValue.cpp [hunk 1]
```diff
@@ -199,5 +200,76 @@ String CSSShapeSegmentValue::customCSSText() const
     return builder.toString();
 }
 
-} // namespace WebCore
 
+BasicShapeShape::ShapeSegment CSSShapeSegmentValue::toShapeSegment(const CSSToLengthConversionData& conversionData) const
```
Comment (weinig) [inline Source/WebCore/css/CSSShapeSegmentValue.cpp]: Generally, I have been preferring to pass Style::BuilderState& for functions like this, and then using Style::BuilderConverter's helpers where possible. That allows a bit more bottlenecking/shared conversion code (which takes things like zoom into account where sometimes it is forgotten).

Can that be done here?

Comment (weinig) [reply Source/WebCore/css/CSSShapeSegmentValue.cpp]: Oh, whoops, I didn't see this was a draft, sorry.

File: Source/WebCore/rendering/style/BasicShapesShape.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (C) 2012 Adobe Systems Incorporated. All rights reserved.
```
Comment (nt1m) [inline Source/WebCore/rendering/style/BasicShapesShape.cpp]: Is this correct? Ditto for some other files

File: LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-computed.tentative.html
Comment (nt1m) [inline LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-computed.tentative.html:1]: Missing -expected.txt files

File: LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-valid.tentative.html [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+<script src="/css/support/parsing-testcommon.js"></script>
+<style>
+    #target {
+        width: 200px;
+        height: 100px;
+    }
+</style>
+</head>
+<body>
+<div id="target"></div>
```
Comment (nt1m) [inline LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-valid.tentative.html]: You shouldn't need the `#target` element for the -valid.html test. Only the -computed.html really needs it.

**Review by nt1m — APPROVED** [review]

File: Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp [hunk 4]
```diff
@@ -2606,6 +2608,217 @@ static RefPtr<CSSPathValue> consumeBasicShapePath(CSSParserTokenRange& args, Opt
...
+
+        return token == CSSValueBy ? CoordinateAffinity::Relative : CoordinateAffinity::Absolute;
+    };
+
+    auto atEndOfCommand = [&] () {
+        return range.atEnd() || range.peek().type() == CommaToken;
+    };
+
+    auto id = range.consumeIncludingWhitespace().id();
+    if (id == CSSValueMove) {
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp]: Can you do a switch statement?

File: Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp [hunk 5]
```diff
@@ -2606,6 +2608,217 @@ static RefPtr<CSSPathValue> consumeBasicShapePath(CSSParserTokenRange& args, Opt
...
+}
+
+// https://drafts.csswg.org/css-shapes-2/#shape-function
+static RefPtr<CSSShapeValue> consumeBasicShapeShape(CSSParserTokenRange& range, const CSSParserContext& context, OptionSet<PathParsingOption> options)
+{
+    if (!context.cssShapeFunctionEnabled)
+        return nullptr;
+
+    // shape() = shape( <'fill-rule'>? from <coordinate-pair>, <shape-command>#)
+    WindRule rule = WindRule::NonZero;
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp]:
```suggestion
    auto rule = WindRule::NonZero;
```

File: Source/WebCore/css/BasicShapesShapeSegmentConversion.cpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                segment.arcSize() == ShapeArcSegment::ArcSize::Large ? CSSValueLarge : CSSValueSmall,
+                CSSPrimitiveValue::create(segment.angle(), CSSUnitType::CSS_DEG));
+        },
+        [&](const ShapeCloseSegment&) {
+            return CSSShapeSegmentValue::createClose();
+        }
+    );
+}
+
+#pragma mark -
```
Comment (nt1m) [inline Source/WebCore/css/BasicShapesShapeSegmentConversion.cpp]:
```suggestion
# MARK -
```
?

File: Source/WebCore/css/BasicShapeConversion.cpp [hunk 7]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (C) 2012 Adobe Systems Incorporated. All rights reserved.
+ * Copyright (C) 2042 Apple Inc. All rights reserved.
```
Comment (nt1m) [inline Source/WebCore/css/BasicShapeConversion.cpp]: 2042

File: Source/WebCore/css/CSSShapeSegmentValue.cpp (hunk 1 above)
Comment (smfr) [reply Source/WebCore/css/CSSShapeSegmentValue.cpp]: I did this, and it caused an SVG zoom bug which I fixed.

**Pull request merged**